File: backend/app/models/match.py
# -*- coding: utf-8 -*-
"""
Pydantic models for match-related operations, mirroring the Matches database schema.
"""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field

from app.models.team import TeamResponse


class MatchBase(BaseModel):
    """
    Base model for match properties directly corresponding to schema.

    Attributes
    ----------
    winner_team_id : int
        ID of the winning team.
    loser_team_id : int
        ID of the losing team.
    is_fanny : bool, optional
        Whether the match was a 'fanny' (default is False).
    played_at : datetime
        Timestamp when the match was played.
    """

    winner_team_id: int = Field(..., gt=0, description="ID of the winning team")
    loser_team_id: int = Field(..., gt=0, description="ID of the losing team")
    is_fanny: bool = Field(default=False, description="Whether the match was a 'fanny'")
    played_at: datetime = Field(..., description="Timestamp when the match was played")
    notes: Optional[str] = Field(default=None, description="Optional notes about the match")


class MatchCreate(MatchBase):
    """
    Data model for creating a new match record.
    """

    pass


class MatchUpdate(BaseModel):
    """
    Data model for updating an existing match's information.
    Matches are typically immutable once recorded. Placeholder for future needs.
    """

    pass


class MatchResponse(MatchBase):
    """
    Data model for returning match information in API responses.

    Attributes
    ----------
    match_id : int
        Unique identifier for the match.
    winner_team : Optional[TeamResponse]
        Detailed information for the winning team (populated by service layer).
    loser_team : Optional[TeamResponse]
        Detailed information for the losing team (populated by service layer).
    """

    match_id: int = Field(..., gt=0, description="Unique identifier for the match")
    winner_team: Optional[TeamResponse] = Field(default=None, description="Details for the winning team")
    loser_team: Optional[TeamResponse] = Field(default=None, description="Details for the losing team")
    notes: Optional[str] = Field(default=None, description="Optional notes about the match")

    class Config:
        from_attributes = True


class MatchWithEloResponse(MatchResponse):
    """
    Extended match response that includes ELO changes.

    Attributes
    ----------
    elo_changes : dict
        Dictionary mapping player IDs to their ELO changes.
    """

    elo_changes: dict = Field(default_factory=dict, description="ELO changes for each player")

File: lib/types/schemas/match.ts
import { z } from "zod";
import { TeamResponseSchema } from "./team";

export const MatchBaseSchema = z.object({
  winner_team_id: z.number().int().positive(),
  loser_team_id: z.number().int().positive(),
  is_fanny: z.boolean().default(false),
  played_at: z.string().datetime(),
  notes: z.string().nullable().optional(),
});

export const MatchCreateSchema = MatchBaseSchema;

export const MatchUpdateSchema = z.object({});

export const MatchResponseSchema = MatchBaseSchema.extend({
  match_id: z.number().int().positive(),
  winner_team: TeamResponseSchema.optional().nullable(),
  loser_team: TeamResponseSchema.optional().nullable(),
});

// [>]: EloChange structure for elo_changes map.

export const EloChangeSchema = z.object({
  old_elo: z.number().int().nonnegative(),
  new_elo: z.number().int().nonnegative(),
  difference: z.number().int(),
});

export const MatchWithEloResponseSchema = MatchResponseSchema.extend({
  elo_changes: z.record(z.string(), EloChangeSchema).default({}),
});

export type MatchBase = z.infer<typeof MatchBaseSchema>;
export type MatchCreate = z.infer<typeof MatchCreateSchema>;
export type MatchUpdate = z.infer<typeof MatchUpdateSchema>;
export type MatchResponse = z.infer<typeof MatchResponseSchema>;
export type EloChange = z.infer<typeof EloChangeSchema>;
export type MatchWithEloResponse = z.infer<typeof MatchWithEloResponseSchema>;
